File: ios/LocaleUnits.swift
import Foundation

// ── LocaleUnits ───────────────────────────────────────
// Pure locale resolution + metric/imperial unit derivation extracted from
// `HybridReactNativeMapboxNavigation.ensureMapboxNavigation`. Operates on a
// plain optional `String` language code so it is unit-testable without the
// Nitro runtime.
//
// The SDK ultimately derives the spoken/displayed measurement system from
// the `CoreConfig.locale`; `usesMetric(for:)` exposes that same decision as
// a pure function so it can be asserted in tests and mirrored on Android.
enum LocaleUnits {

  /// Resolve the consumer's optional `language` prop into a `Locale`.
  /// A non-empty BCP-47 code is honoured verbatim; otherwise we fall back to
  /// the device's nationalised current locale (matching the previous inline
  /// behaviour exactly).
  static func resolveLocale(language: String?) -> Locale {
    if let code = language, !code.isEmpty {
      return Locale(identifier: code)
    }
    return Locale(identifier: nationalizedCurrentIdentifier())
  }

  /// Pure-Foundation replica of `MapboxNavigationCore`'s
  /// `Locale.nationalizedCurrent` (language from the app's preferred
  /// localization, region from the current locale). Inlined so this unit
  /// stays SDK-free and host-testable instead of importing the SDK for one
  /// extension.
  static func nationalizedCurrentIdentifier() -> String {
    let first = Bundle.main.preferredLocalizations.first ?? Locale.current.identifier
    let parts = first.components(separatedBy: "-")
    if parts.count > 1 { return first }
    if let countryCode = (Locale.current as NSLocale).object(forKey: .countryCode) as? String {
      return "\(parts.first ?? first)-\(countryCode)"
    }
    return first
  }

  /// Whether the given locale uses the metric measurement system. This is
  /// the derivation Mapbox applies internally from the locale; exposed here
  /// as a pure, testable seam. Defaults to metric when the platform can't
  /// determine the system (the global-majority default).
  static func usesMetric(for locale: Locale) -> Bool {
    if #available(iOS 16, macOS 13, *) {
      switch locale.measurementSystem {
      case .us, .uk:
        return false
      default:
        return true
      }
    } else {
      return locale.usesMetricSystem
    }
  }
}

File: android/src/main/java/com/margelo/nitro/jlsdigital/reactnativemapboxnavigation/LocaleUnits.kt
package com.margelo.nitro.jlsdigital.reactnativemapboxnavigation

import com.mapbox.navigation.base.formatter.UnitType
import java.util.Locale

// ── LocaleUnits ───────────────────────────────────────
// Pure locale resolution + metric/imperial unit derivation extracted from
// `HybridReactNativeMapboxNavigation` (the inline `resolveLocale` /
// `unitTypeFor`). Operates on a plain optional `String` language code and a
// `Locale` so it is unit-testable without the Nitro runtime; the only SDK
// touch is the value enum `UnitType`, which is a plain JVM enum.
//
// Mirrors iOS `LocaleUnits` (resolveLocale + usesMetric).
object LocaleUnits {

  /**
   * Resolve the consumer's optional `language` prop into a [Locale].
   * A non-blank tag is honoured verbatim (with `_` normalised to `-` for
   * BCP-47); otherwise we fall back to the device's default locale (matching
   * the previous inline behaviour exactly).
   */
  fun resolveLocale(language: String?): Locale {
    val tag = language?.takeIf { it.isNotBlank() } ?: return Locale.getDefault()
    return Locale.forLanguageTag(tag.replace('_', '-'))
  }

  /**
   * Whether the given locale uses imperial or metric units. Mirrors the
   * country-code derivation the god class applied inline: the imperial
   * holdouts (US, Liberia, Myanmar) get [UnitType.IMPERIAL], everything else
   * is [UnitType.METRIC].
   */
  fun unitTypeFor(locale: Locale): UnitType = when (locale.country.uppercase()) {
    "US", "LR", "MM" -> UnitType.IMPERIAL
    else -> UnitType.METRIC
  }
}
